### PR TITLE
[Core] Improve MetaTensorMode to intercept more tensor factory operations

### PR DESCRIPTION
**Pull request by calleum**
## Purpose

This PR fixes GPU memory allocation issues during model initialization for tensorizer by expanding `MetaTensorMode`'s coverage from 1 to 18 tensor factory operations.

- Fixes #25751 - tensorizer model loading with tensor parallelism increases memory requirements (GPU memory doubling from ~35GB to ~70GB per GPU with Llama-3-70B TP=4)
- Fixes #32925 - tensorize_vllm_model double gpu (memory usage spiking to 32GB then dropping to 21GB during initialization)

Both issues were later stale-closed rather than fixed. `MetaTensorMode` on main is unchanged.

Previously, only `aten::empty` was intercepted, causing operations like torch.zeros(), torch.ones(), torch.rand(), etc. to allocate GPU memory during model initialization.

Changes:
- Add `_FACTORY_OPS` frozenset with 18 tensor factory operations
- Force meta device even when device is explicitly specified
- Use safer kwargs handling (explicit None check)
- Add documentation to MetaTensorMode class

This change fixes incorrect GPU memory allocation during model initialization and makes tensorizer more robust for large models and multi-GPU setups.

## Test Plan
Run new tests after loading venvs - some are failing already, but not affected by my changes 
```sh
pytest tests/model_executor/model_loader/tensorizer_loader/
```

## Test Result
```sh
platform linux -- Python 3.12.12, pytest-9.0.2, pluggy-1.6.0 -- /home/tiny/src/vllm/.venv/bin/python
     cachedir: .pytest_cache
     rootdir: /home/tiny/src/vllm
     configfile: pyproject.toml
     plugins: asyncio-1.3.0, anyio-4.12.1
     asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
     collecting ... collected 9 items

     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_basic_factory_ops PASSED [ 11%]
     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_like_operations PASSED [ 22%]
     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_strided_operations PASSED [ 33%]
     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_device_override_cpu PASSED [ 44%]
     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_device_override_cuda PASSED [ 55%]
     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_new_tensor_methods PASSED [ 66%]
     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_kwargs_none_handling PASSED [ 77%]
     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_factory_ops_coverage PASSED [ 88%]
     tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py::TestMetaTensorMode::test_outside_context_manager PASSED [100%]

     =============================== warnings summary ===============================
     <frozen importlib._bootstrap>:488
       <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

     <frozen importlib._bootstrap>:488
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/model_loader/tensorizer.py [hunk 1]
```diff
@@ -87,10 +87,43 @@ def tensorizer_kwargs_arg(value):
...
+            "aten::new_zeros",
+            "aten::new_ones",
+            "aten::new_full",
+            "aten::new_empty_strided",
+        }
+    )
+
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
-        kwargs = kwargs or {}
+        kwargs = kwargs if kwargs is not None else {}
```
Comment (sangstar) [inline vllm/model_executor/model_loader/tensorizer.py]: This is not a meaningful `None` check even though it explicitly checks for it. The original implicitly checks for `None` as `None` is falsy. This change just preserves falsy types that `kwargs` shouldn't be (like `""`, `0`, `False` etc) which would immediately cause an error when the type accesses `dict` methods.


Really, if we're going to be checking `kwargs` we ought to actually be enforcing type correctness if it's not done elsewhere, as neither version does so.

File: vllm/model_executor/model_loader/tensorizer.py [hunk 2]
```diff
@@ -87,10 +87,43 @@ def tensorizer_kwargs_arg(value):
...
 
-        if func._schema.name == "aten::empty" and "device" not in kwargs:
+        if func._schema.name in self._FACTORY_OPS:
+            # Force device to meta, overriding any specified device
```
Comment (sangstar) [inline vllm/model_executor/model_loader/tensorizer.py:126]: The override here is an important fix. 👍

**Review by sangstar — COMMENTED** [review]
Thanks for this great change! Slight nitpick on the type safety of `kwargs`. Great tests as well.

File: vllm/model_executor/model_loader/tensorizer.py (hunk 1 above)
Comment (calleum) [reply vllm/model_executor/model_loader/tensorizer.py]: Yeah, you're right. They're equivalent under the dispatch protocol anyway. Reverted in 98c0e114b.

**Comment by calleum** [comment]
@sangstar tests are gated behind the `ready` label for first-time contributors. Could you add it when you get a chance? The kwargs revert is in 98c0e114b.

File: vllm/model_executor/model_loader/tensorizer.py [hunk 3]
```diff
@@ -87,10 +87,43 @@ def tensorizer_kwargs_arg(value):
...
+            "aten::rand_like",
+            "aten::randn_like",
+            "aten::empty_strided",
+            "aten::new_empty",
+            "aten::new_zeros",
+            "aten::new_ones",
+            "aten::new_full",
+            "aten::new_empty_strided",
+        }
+    )
```
Comment (OPS-NeoRetro) [inline vllm/model_executor/model_loader/tensorizer.py:120]: Good. torch.empty, torch.zeros, torch.ones, torch.full, torch.randn, torch.zeros_like, torch.new_zeros and other similar functions are properly intercepted.

File: vllm/model_executor/model_loader/tensorizer.py [hunk 4]
```diff
@@ -87,10 +87,43 @@ def tensorizer_kwargs_arg(value):
...
+
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         kwargs = kwargs or {}
 
-        if func._schema.name == "aten::empty" and "device" not in kwargs:
+        if func._schema.name in self._FACTORY_OPS:
```
Comment (OPS-NeoRetro) [inline vllm/model_executor/model_loader/tensorizer.py:125]:
```suggestion
        if func._schema.name in self._FACTORY_OPS and "device" not in kwargs:
```

Why did you delete `and "device" not in kwargs`?

File: tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        """Test that strided operations create meta tensors."""
+        with MetaTensorMode():
+            tensor = torch.empty_strided((10, 20), (20, 1))
+            assert tensor.device.type == "meta"
+
+    def test_device_override_cpu(self):
+        """Test that explicit CPU device is overridden to meta."""
+        with MetaTensorMode():
+            tensor = torch.empty(10, 20, device="cpu")
+            assert tensor.device.type == "meta"
```
Comment (OPS-NeoRetro) [inline tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py:50]: Is this the intended behaviour?

File: tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+    def test_device_override_cuda(self):
+        """Test that explicit CUDA device is overridden to meta."""
+        with MetaTensorMode():
+            tensor = torch.empty(10, 20, device="cuda")
+            assert tensor.device.type == "meta"
```
Comment (OPS-NeoRetro) [inline tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py:57]: Is this also the intended behaviour?

**Review by OPS-NeoRetro — CHANGES_REQUESTED** [review]
This is the second code review on this stale PR. @calleum why did you or your AI agent make those oversights? Are those intentional or not?

**Comment by OPS-NeoRetro** [comment]
Bump @jikunshang @sangstar @robertgshaw2-redhat @bigPYJ1151

**Comment by OPS-NeoRetro** [comment]
@calleum I did a second code review a week ago

File: vllm/model_executor/model_loader/tensorizer.py (hunk 4 above)
Comment (calleum) [reply vllm/model_executor/model_loader/tensorizer.py:125]: intentional. the doubling comes from modules that allocate with an explicit device in __init__. @sangstar flagged the same line in march: https://github.com/vllm-project/vllm/pull/33235#discussion_r2995809522

File: tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py (hunk 5 above)
Comment (calleum) [reply tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py:50]: yep, intended. an explicit device has to be overridden or the allocation still happens.

File: tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py (hunk 6 above)
Comment (calleum) [reply tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py:57]: same, intended. explicit cuda is the cause of the related issues

**Comment by calleum** [comment]
thanks @OPS-NeoRetro, answered inline. the override is intentional. updated comments in the code, your flags made me notice they were out of sync with the actual code

File: tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py [hunk 7]
```diff
@@ -87,33 +86,32 @@ def test_factory_ops_coverage(self):
...
             assert torch.zeros_like(ref_tensor).device.type == "meta"
             assert torch.ones_like(ref_tensor).device.type == "meta"
             assert torch.full_like(ref_tensor, 2.0).device.type == "meta"
             assert torch.rand_like(ref_tensor).device.type == "meta"
             assert torch.randn_like(ref_tensor).device.type == "meta"
 
-            # Strided operation (1)
+            # Strided operations
             assert torch.empty_strided((5, 5), (5, 1)).device.type == "meta"
+            assert ref_tensor.new_empty_strided((5, 5), (5, 1)).device.type == "meta"
```
Comment (OPS-NeoRetro) [inline tests/model_executor/model_loader/tensorizer_loader/test_meta_tensor_mode.py:107]: I didn't notice that this was not tested at all.

**Review by OPS-NeoRetro — APPROVED** [review]
I'm now approving this PR. Waiting for vLLM maintainers to look at this PR one more time.

**Comment by OPS-NeoRetro** [comment]
@calleum I need you to retest this PR merged into your fork of the nightly version of vLLM. The issues linked to this PR that were closed as unplanned were unreproducible bugs, and https://github.com/intel/llm-scaler/issues/578 was also intentionally closed because the person who opened the issue can't reproduce it. These allocation problems actually stem from PyTorch not properly freeing VRAM after exiting or something when they tried out multiple local models one at a time, or even Wayland, or some Skia-based application like an Electron, desktop Flutter or pixel-prefect Uno Platform app not properly freeing VRAM. Those unreproducible vLLM bugs only happen on Linux, which means something has to do with Linux or Linux distros, or PyTorch (which is the base of vLLM) with CUDA and SYCL on Linux. I'm trying to know what happened before those unreproducible tensor parallelizer bugs happened. Maybe it's because of background apps, they ran some VRAM-heavy or RAM-heavy workload before or simply because some GPU drivers have shared VRAM oversights, or something else related to VRAM or RAM management. Even the comment closing that iPEx-related issue said this.

> So uhhh I can't reproduce this anymore either. I don't know what I did. Maybe some obscure driver memory allocation bug that was fixed by a reboot?

**Comment by calleum** [comment]
Ping @sangstar

**Comment by calleum** [comment]
Bump @sangstar

**Review by sangstar — APPROVED** [review]
I'm satisfied with these changes now, but let's make sure the comment [here](https://github.com/vllm-project/vllm/pull/33235#issuecomment-5201690178) has been addressed before this gets merged.

**Comment by calleum** [comment]
> @calleum I need you to retest this PR merged into your fork of the nightly version of vLLM. The issues linked to this PR that were closed as unplanned were unreproducible bugs, and https://github.com/intel/llm-scaler/issues/578 was also intentionally closed because the person who opened the issue can't reproduce it. These allocation problems actually stem from PyTorch not properly freeing VRAM after exiting or something when they tried out multiple local models one at a time, or even Wayland, or some Skia-based application like an Electron, desktop Flutter or pixel-prefect Uno Platform app not properly freeing VRAM. Those unreproducible vLLM bugs only happen on Linux, which means something has to do with Linux or Linux distros, or PyTorch (which is the base of vLLM) with CUDA and SYCL on Linux. I'm trying to know what happened before those unreproducible tensor parallelizer bugs happened. Maybe it's because of background apps, they ran some VRAM-heavy or RAM-heavy workload before or simply because some GPU drivers have shared VRAM oversights, or something else related to VRAM or RAM management. Even the comment closing that iPEx-related issue said this.
> 
> > So uhhh I can't reproduce this anymore either. I don't know what I did. Maybe some obscure driver memory allocation bug that was fixed by a reboot?


I am still able to reproduce my issue and validate my commit fixes it. 

The issues in vllm that this pr links to were closed for inactivity, not because they cannot be reproduced.
